Rewrite ffe-radio-button-react

selbekk opened the pull request:
This commit rewrites the `ffe-radio-button-react` implementation to improve, simplify and modernize the API.

This is an initial pull request - it's not complete, but complete enough to get some feedback.

A quick summary of this PR is the following:

There are now four components:
- `<RadioButton />` - a simple radio button choice
- `<RadioSwitch />` - a binary radio button choice
- `<RadioButtonGroup />` - a wrapping component simplifying the use of the previous two
- `<RadioBlock />` <-- this guy is new - well, for React, at least.

There's another internal component called `<BaseRadioButton />` that is the base unit creating the three first components.

Note that the `<RadioButtonGroup />` is requiring a [render-prop (or function-as-a-child)](https://reactjs.org/docs/render-props.html) - this is something that's not heavily used in FFE yet. It _is_ used by `ffe-form-react`'s `<InputGroup />` component, and it works just fine. It's a different, less magical way of intuitively passing props to children.

Let me know what you think!

Fixes #50

wkillerud commented on packages/ffe-radio-button-react/src/RadioButton.md:
````diff
@@ -0,0 +1,88 @@
+</RadioButtonGroup>
+```
+
+Noen valg kan være mer kompliserte, og ikke like selvforklarende. Da kan man
````
Is this really by design, or is the tooltip there for historical reasons? Seems like an odd place to explain complex options - I'd think a tooltip explaining the whole RadioButtonGroup would be better.

selbekk replied:
Probably historical. I guess we can ask the designers! 

@adidick - should each radio button choice be able to have its own tooltip? With today's implementation, it looks a bit silly.

adidick-zz replied:
I'm not sure exactly what you mean here @selbekk. Can we take a look at it together when I'm back (currently home with sick kids)?

selbekk replied:
@adidick This is what we're talking about - per-radio button tooltips:

![image](https://user-images.githubusercontent.com/1307267/37057273-1f660100-2187-11e8-95d3-f8de4d57e354.png)

Should this be allowed at all, or should we just have them per group - like this?

![image](https://user-images.githubusercontent.com/1307267/37057325-4c319744-2187-11e8-83af-f0d67eea8429.png)

Hope your kids feel better soon!

wkillerud commented on packages/ffe-radio-button-react/src/RadioBlock.js:
```diff
@@ -0,0 +1,64 @@
+        const isSelected = checked || selectedValue === value;
+
+        return (
+            <div className="ffe-radio-block">
```
Is a `className` prop relevant here?

selbekk replied:
Yep! Nice catch. We might need that for some of the child elements as well, but I think I'll leave that part until we actually DO need them.

wkillerud commented on packages/ffe-radio-button-react/src/RadioButtonGroup.js:
```diff
@@ -0,0 +1,80 @@
+                        }
+                    </legend>
+                }
+                {children(buttonProps)}
```
I approve this pattern :rocket: So much more flexible for wrappers like this.

selbekk replied:
I mentioned this in a [thread on Twitter](https://twitter.com/selbekk/status/969960538003427328) to Kent Dodds - the guy that popularized this render prop pattern. He said

> If the state you want to share is implicit then compound components are great. Otherwise render props are much more powerful and flexible.

I think we want render props for this. The [react docs even have their own section pm them](https://reactjs.org/docs/render-props.html) if you want to learn more.

wkillerud replied:
For consistency could we also have the option for single child use the "old fashioned way" without render prop, [like InputGroup](https://github.com/SpareBank1/designsystem/blob/d6baf4bb982c1bc1e2d98b72f1dac315a4a5d68e/packages/ffe-form-react/src/InputGroup.js#L27)?

selbekk replied:
I'm actually thinking no. In my opinion, allowing for both ways was wasn't a good solution. It doubles the API, which makes it hard to understand (and even harder to explain) when to use what. Sure it's less boilerplate, but it's also more of a "hm what should I do here" and "aww crap I have to switch from one to the other" moment prone technique.

If there are very strong opinions on this I'll budge, but I kind of think this is the way to go.

wkillerud replied:
> It doubles the API, which makes it hard to understand (and even harder to explain) when to use what.

One child - compose as you normally would. Two or more children - use the render prop. I think that's simple enough to explain. Coupled with a helpful error message when trying to compose as normal with two or more children, I don't see the problem as a user. It's worked very well for `InputGroup` in my opinion.

> it's also more of a "hm what should I do here"

Having the thing break when I try to compose as normal with a single child is also a bit of a "hm what should I do here" :smile: 

> it's also more of a (...) "aww crap I have to switch from one to the other" moment prone technique

True, and it's a bit annoying - I'll give you that. Still, I prefer reducing the amount of boilerplate for the `RadioSwitch` case over always using the render prop. It's not a complex rewrite going from one pattern to the other.

wkillerud replied:
@selbekk and I discussed this and he has me convinced that render prop only is the way to go. It avoids the case where a consumer just wraps the two or more children in a `Fragment` or whatever to make the error go away. Less ways to mess things up :+1:

selbekk replied:
I love these discussions! :tada: 

Reality is, there was a lot of confusion when I introduced this new technique for input groups. People did what people do and basically tried to hack around the new API by providing a wrapping DIV (which then received the `aria-invalid` and `id` props, breaking the intended effect of having this component in the first place. :disappointed: 

I agree that it's... nice to have, if you know the API or read the docs extensively, but I think an API with fewer options is better than a "simple-to-use, simple-to-abuse" API. :+1: 

Finally, the only use case for a single child here is when you want a `<RadioSwitch />` in there - and we (or the consumer) can build a wrapper around that use case, simplifying its general use case.

wkillerud commented on packages/ffe-form/less/radio-switch.less:
```diff
@@ -28,6 +28,10 @@
         margin-bottom: 0;
     }
+
+    & + input + .ffe-radio-switch {
```
This change is non-breaking I take it (from the commit message)?

selbekk replied:
Yep that's right. Let me explain: 

In the previous implementation, two radio switch elements (there are typically two of them) had to be separated by inline styles. That was the way this was solved in the existing React implementation at least. This approach solves this issue by a clever CSS selector. 

To make it a bit easier to understand what I'm doing - here's the typical DOM structure of a pair of radio switches:

```html
<input id="yes" type="radio" class="ffe-radio-input" />
<label for="yes" class="ffe-radio-switch">Yes</label>

<input id="no" type="radio" class="ffe-radio-input" />
<label for="no class="ffe-radio-switch">No</label>
```

I guess I could have used the class instead, but in reality that doesn't matter much.

wkillerud reviewed (commented):
Liking the changes to the `RadioButtonGroup`, and it'll be nice to have a `Group`-less `RadioSwitch` :+1:

wkillerud commented on packages/ffe-radio-button-react/src/RadioButtonGroup.js:
```diff
@@ -0,0 +1,80 @@
+        return (
+            <fieldset 
+                className="ffe-fieldset" 
+                onChange={this.onChange}
```
This should be `onChange` (sans `this`), right? To use the prop. With that this can also be a stateless functional component.

selbekk replied:
Very true - this was some left-behind garbage from a previous iteration. Nicely spotted!

wkillerud commented on packages/ffe-radio-button-react/src/RadioSwitch.js:
```diff
@@ -0,0 +1,42 @@
+    return (
+        <Fragment>
+            <BaseRadioButton
+                className="ffe-radio-switch"
```
We need the same `classNames` handling here as in `RadioButton`, otherwise a `className` set on the component will override the FFE class.

selbekk replied:
I agree!

wkillerud commented on packages/ffe-radio-button-react/src/RadioButtonGroup.js:
```diff
@@ -0,0 +1,80 @@
+
+        return (
+            <fieldset 
+                className="ffe-fieldset" 
```
I'd love a `className` prop here as well :place_of_worship:

selbekk replied:
No problem mister!

selbekk commented:
@wkillerud is killing it (wink wink) with the review comments here - anyone else interested in this? cc @SpareBank1/designsystem  @SpareBank1/designsystem-collaborators

adidick-zz commented:
It’s hard to give a concrete answer without knowing if there are any cases where we might need one info button on each radio button. My gut tells me that that’s overkill and that we should be doing more to simplify the possible answers by presenting all the info up front so that when it comes to choosing the user can do it easily with a full understanding of the question.

Sent from my iPhone

> On 7/03/2018, at 8:19 AM, William Killerud <notifications@github.com> wrote:
> 
> @wkillerud commented on this pull request.
> 
> In packages/ffe-radio-button-react/src/RadioButtonGroup.js:
> 
> > +        };
> +
> +        return (
> +            <fieldset 
> +                className="ffe-fieldset" 
> +                onChange={this.onChange}
> +            >
> +                {description && 
> +                    <legend className="ffe-form-label">
> +                        {description}
> +                        {tooltip &&
> +                            <Tooltip {...tooltipProps}>{tooltip}</Tooltip>
> +                        }
> +                    </legend>
> +                }
> +                {children(buttonProps)}
> It doubles the API, which makes it hard to understand (and even harder to explain) when to use what.
> 
> One child - compose as you normally would. Two or more children - use the render prop. I think that's simple enough to explain. Coupled with a helpful error message when trying to compose as normal with two or more children, I don't see the problem as a user. It's worked very well for InputGroup in my opinion.
> 
> it's also more of a "hm what should I do here"
> 
> Having the thing break when I try to compose as normal with a single child is also a bit of a "hm what should I do here" 😄
> 
> it's also more of a (...) "aww crap I have to switch from one to the other" moment prone technique
> 
> True, and it's a bit annoying - I'll give you that. Still, I prefer reducing the amount of boilerplate for the RadioSwitch case over always using the render prop. It's not a complex rewrite going from one pattern to the other.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

selbekk commented:
Sounds like a face-to-face discussion sometime later this week @adidick :slightly_smiling_face: We'll discuss this later this week.

selbekk committed:
chore: Add support for class property inits

This commit adds support for class property initializers, which lets
us write simpler code. It's a stable spec ([stage 3](https://github.com/tc39/proposal-class-fields)), so it should be ready to start using.

selbekk committed:
fix(ffe-form): Add space between radio switches

This commit adds some margin between radio switch components when
they are stacked next to each other.

selbekk committed:
feat(ffe-radio-button-react): Reimplement

BREAKING CHANGE: This commit reimplements the
`ffe-radio-button-react` package, and modernizes its APIs.

This change removes deprecated APIs, and instead takes a fresh approach
to improve the developer experience and design consistency.

Here is a non-exhaustive list of breaking changes:

*`<RadioButton />`*
- `aria-invalid` prop must be `'true'` or `'false'` now - no booleans!
- `invalid` prop is removed - use `aria-invalid`

*`<RadioSwitch />`*
- The `<RadioSwitch />` is no longer wrapped in a `RadioButtonGroup`
component! Wrap it yourself - create a wrapping component if it's a lot
of work for your usecase.
- `aria-invalid` prop must be `'true'` or `'false'` now - no booleans!
- `invalid` prop is removed - use `aria-invalid`
- `leftOption` object is now replaced by two props - `leftLabel` and
`leftValue`
- `rightOption` object is now replaced by two props - `rightLabel` and
`rightValue`.

*`<RadioButtonInputGroup />`*
- This component now requires you to pass a function as a child. You can
[read more about this pattern here](https://reactjs.org/docs/render-props.html).
The function receives an object with all props relevant to all options.
Spread these to every radio button.
- `buttons` prop is deprecated - pass your choices in as children
- `aria-invalid` prop must be `'true'` or `'false'` now - no booleans!
- `invalid` prop is removed - use `aria-invalid`

Please refer to the documentation for usage examples.

Finally, this commit documents this package, which had previously
been forgotten about.

wkillerud reviewed (dismissed):
Nice work! :tada:
